Add Const Generic Image Type

XAMPPRocky opened the pull request:
This PR refactors `Image` in the backend to not use attributes to determine things like dimensionality and depth, and to instead accept them as const generics as parameters. This has the advantage of only needing a single `Image` type as opposed previously to requiring a new type for each combination, and also allows us to have enums to represent all of the possibilities rather than using `u32`s directly.

### Additions

- `macros::Image` a new procedural macro for defining `Image` types.
- `image::SampleType` a trait that allows us to be generic of the underlying sampled type of the image.
- `image::ImageCoordinate` a trait built on top of `Vector` for parameterising image coordinate vectors based on the sampled type, dimensionality, and arrayed-ness of the Image.
- A new `spirv_types` crate for holding types that are useful in both `spirv-std` and `spirv-std-macros`.

### Re-works

- `crate/spirv-std-macros` has been moved into the `spirv-std` folder as `macros` beside the `shared` crate, this keeps the folder structure cleaner.

expenses commented:
Awesome! I'm glad validation catches things like sampling with the wrong coordinate type:
```rust
error: invalid data:0:0 - Expected Coordinate to have at least 3 components, but given only 1
  %85 = OpImageSampleImplicitLod %v4float %84 %77

  |
  = note: spirv-val failed
```

I think that type aliases would be really useful though, e.g. 
```rust
pub type ImageCube = Image<
    { Dimensionality::Cube },
    { ImageDepth::No },
    false,
    false,
    { Sampled::Yes },
    { ImageFormat::Unknown },
    { None },
>;
```
or possibly 
```rust
pub type ImageCube<
    const FORMAT: ImageFormat,
> = Image<
    { Dimensionality::Cube },
    { ImageDepth::No },
    false,
    false,
    { Sampled::Yes },
    { FORMAT },
    { None },
>;
```

if there's a performance gain with setting the image format.

XAMPPRocky commented:
> Awesome! I'm glad validation catches things like sampling with the wrong coordinate type:

Yeah, we should probably see if we can catch that at a higher level so that you see the `sample` call from the Rust code rather than the SPIR-V asm. 

> I think that type aliases would be really useful though

Sure, I don't really see the harm in that, though that's really just working around the current limitations of the lack of defaults for const generics. Once defaults are implemented we could set them on `Image` itself so that most users could just say `Image<{ Dimensionality::Cube }>` in most cases, and then if users need a more complex combination they can set their own type aliases.

eddyb commented on crates/spirv-std/src/textures.rs:
```diff
@@ -1,10 +1,81 @@
+    Rg16f = 7,
+    R11fG11fB10f = 8,
+    R16f = 9,
+    Rgba16 = 10,
+    Rgb10A2 = 11,
+    Rg16 = 12,
+    Rg8 = 13,
+    R16 = 14,
+    R8 = 15,
+    Rgba16Snorm = 16,
+    Rg16Snorm = 17,
+    Rg8Snorm = 18,
+    R16Snorm = 19,
+    R8Snorm = 20,
+    Rgba32i = 21,
+    Rgba16i = 22,
+    Rgba8i = 23,
+    R32i = 24,
+    Rg32i = 25,
+    Rg16i = 26,
+    Rg8i = 27,
+    R16i = 28,
+    R8i = 29,
+    Rgba32ui = 30,
+    Rgba16ui = 31,
+    Rgba8ui = 32,
+    R32ui = 33,
+    Rgb10a2ui = 34,
+    Rg32ui = 35,
+    Rg16ui = 36,
+    Rg8ui = 37,
+    R16ui = 38,
+    R8ui = 39,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum AccessQualifier {
+    ReadOnly = 0,
+    WriteOnly = 1,
+    ReadWrite = 2,
```
Rust guarantees sequential values starting from `0`, not sure why they were explicitly provided here.

XAMPPRocky replied:
Well these are also the values of their SPIR-V assembly, and including them would allow us to re-order them and not worry about it affecting code, where as implicit sequential values relies on us, rustfmt, or a contributor remembering to not re-order them.

eddyb commented on crates/spirv-std/src/textures.rs:
```diff
@@ -52,58 +150,41 @@ impl Image2d {
+        >,
+    >
+{
+    pub fn sample(&self, coord: impl ScalarOrVector) -> Vec4 {
```
Shouldn't the coordinate type depend on the dimensionality of the image?

You can replace `trait ScalarOrVector` with a `trait ImageCoord<const DIM: Dimensionality>` you can implement for specific dimensionalities, though it might make even more sense to extract a single coordinate type, instead.

Though with a `ImageCoord` trait, we can support both `Vec3` and `Vec3A`, by converting the former to the latter.

XAMPPRocky replied:
> Shouldn't the coordinate type depend on the dimensionality of the image?

Yes it should, just wasn't sure how to ensure that. That idea sounds great and I'll add it.

XAMPPRocky commented on crates/spirv-std/src/textures.rs:
```diff
@@ -18,18 +89,45 @@ pub struct Sampler {
+        const ACCESS_QUALIFIER: Option<AccessQualifier>,
+    >
+    Image<
+        { DIM },
+        { DEPTH },
+        { ARRAYED },
+        { MULTISAMPLED },
+        { SAMPLED },
+        { FORMAT },
+        { ACCESS_QUALIFIER },
```
They're currently required for ADT types AFAICT. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=47f4ed26b3f3965e164d075aede833fa

eddyb replied:
Name resolution doesn't care about types. These are generic parameter names, so they should work just fine unbraced.

Jasper-Bekkers commented on examples/shaders/sky-shader/src/lib.rs:
```diff
@@ -159,14 +160,29 @@ pub fn fs(constants: &ShaderConstants, frag_coord: Vec2) -> Vec4 {
+    #[spirv(frag_coord)] in_frag_coord: Input<Vec2>,
     constants: PushConstant<ShaderConstants>,
+    #[spirv(binding = 0)] img: UniformConstant<
+        Image<
+            { Dimensionality::TwoD },
+            { ImageDepth::No },
+            false,
+            false,
+            { Sampled::Yes },
+            { ImageFormat::Unknown },
+            { None },
+        >,
```
I think we can't really have user facing code look like this, it's way to verbose so at a minimum we should have some typedefs for this.

Jasper-Bekkers replied:
Thanks for adding the type aliasses @XAMPPRocky, we should use them here as well, code samples get copy & pasted far and wide.

XAMPPRocky replied:
Yeah, I'll update it, I have to make the example work first. 😉

eddyb commented on rust-toolchain:
```diff
@@ -5,5 +5,5 @@
 
 [toolchain]
-channel = "nightly-2020-12-11"
+channel = "nightly-2020-12-28"
```
I guess you can rebase on `main` to get rid of this.

Jasper-Bekkers commented on crates/spirv-std/src/textures.rs:
```diff
@@ -1,5 +1,37 @@
+}
+
+basic_image_type! {
+    /// A convenience type alias for a one dimensionsal image.
```
Nit: small typo in dimensional (dimensionsal), happens a few more times below.

XAMPPRocky commented:
This is ready for review, one of the main things to consider before merging is that this will require nightly for `spirv-std` everywhere where as right now you can technically compile `spirv-std` on stable on host platforms.

khyperia commented:
> one of the main things to consider before merging is that this will require nightly for `spirv-std` everywhere

Pinging @VZout on this topic, since they did a bunch of work trying to get `spirv-std` building on stable, so I want them to sign off on this.

VZout commented:
> > one of the main things to consider before merging is that this will require nightly for `spirv-std` everywhere
> 
> Pinging @VZout on this topic, since they did a bunch of work trying to get `spirv-std` building on stable, so I want them to sign off on this.

I'm not the person to ok this. But this is small step back for ark. While it is a good change of course I personally think there is not enough reason to merge this. I prefer to leave the shaders intertwined with ark on stable. Especially if we decide to go with trait based rendering this becomes a big problem. 

Can we hold of on this decision once we have had the more technical meeting with Henrik on the rendering clean up and whether we continue with trait based stuff?

XAMPPRocky commented:
I've also updated this PR so this is no longer the case, it just means now that `Image` is only available when in a shader crates, which should be already be case.

khyperia commented:
> it just means now that `Image` is only available when in a shader crate, which should be already be case.

No, this was an explicit decision and intentional design made to make Image available on the CPU. This is because not having it available on the CPU significantly degrades editor experience out of the box, something that's important to us.

![image](https://user-images.githubusercontent.com/953151/104593379-e6e41d00-566f-11eb-8649-ccdfec1e2e64.png)

If you want to override that decision, that's up to you, I just want to make sure we're aware of the decisions we're making instead of flipping back and forth on some things repeatedly without being aware of history (something that's happened in the past).

XAMPPRocky commented:
> This is because not having it available on the CPU significantly degrades editor experience out of the box

This is a good point, it also means docs are wouldn't be available by default. You could enable both of those back by manually adding a feature flag for example and customising your rust-analyser config, but I don't think that's really ideal.

Fundamentally the stable restriction will require one or more trade-offs in the implementation. I've listed all possibilities that I could think of below and their tradeoffs below.


- Exposing `Image` with enum const generics in the API.
	- **Advantages**
		- Users can't create invalid image types in the type system.
		- We can specialise types in arguments based on these const generics (See `ImageCoord`)
	- **Disadvantages**
	    	- `spirv-std` no longer compiles on stable.
- Keeping `Image` with enum const generics but hidden behind `cfg()`.
	- **Advantages**
		- Same as all of the above.
		- Compiles on stable
	- **Disadvantages**
		- Editor and docs experience worse out of the gate.
- Using type level generics instead of const generics. (E.g. like `num`)
	- **Advantages**
		- Compiles on stable
		- Same benefits as const generics
	- **Disadvantages**
		- A lot more boilerplate to achieve the same goal on both the user and the backend.
- Using only primitives (`u8`, `bool`) const generics.
	- **Advantages**
		- Will be compilable on stable soon.
	- **Disadvantages**
		- Users can easily enter the wrong value
		- The generics become more opaque as to what they're doing.
			-  we can provide `consts` in a module to make it easier.
			- Still doesn't stop someone from writing or seeing `Image<1, 0, false, false, 0, 0, 0>`.
- Continuing with attributes.
	- **Advantages**
		- Compilable on stable.
	- **Disadvantages**
		- Requires a combinatorial explosion of the number types for covering each possible combination.

XAMPPRocky commented:
Actually there's one other possibility that just came to mind, which is using the `RUSTC_BOOTSTRAP` flag in Ark when building the shaders, so that it can use the nightly features while still being on stable.

**Advantages**
- Everything works.

**Disadvantage**
- Opens some ark to instability.

khyperia commented:
Removing myself from reviewers until const generics are stabilized and this can land, because I use https://github.com/pulls/review-requested a lot as my to-do list, and a glaring "review this thing!!" makes me distracted and less likely to review the things I'm supposed to be looking at :s

XAMPPRocky commented:
this will be able to land sooner since it's under a flag, but I still need to do a bit of polish work to get it ready for review so I'll move it back to the draft stage for now.

XAMPPRocky commented on crates/spirv-std/shared/src/image_params.rs:
```diff
@@ -0,0 +1,424 @@
+            "rg8ui" => Self::Rg8ui,
+            "r16ui" => Self::R16ui,
+            "r8ui" => Self::R8ui,
+            _ => return Err("Unknown specified image format. Use `unknown(<type>)` if this is intentional."),
```
```suggestion
            _ => return Err("Unknown specified image format. Use `type=<type>` instead if this is intentional."),
```

XAMPPRocky commented:
Actually, need to figure out some more stuff before it's ready for review.

eddyb commented:
The overall approach seems good to me, but I'm not so sure about the crate organization (and the new crate).

It might even make sense to generate some of the bigger `enum`s from e.g. [`spirv.core.grammar.json`](https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/unified1/spirv.core.grammar.json), instead of having them laid out by hand.

I'm deferring to @khyperia on the issue of the (`macros` + `shared`) crate organization, I don't want to block this (long-awaited) PR if she's fine with it.

XAMPPRocky commented:
> It might even make sense to generate some of the bigger enums from e.g. spirv.core.grammar.json, instead of having them laid out by hand.

I originally did it by hand because the definitions are already generated in the `spirv_headers` crate, but that's didn't used to work when I first tired to use it (back in January), but now it seems like it compiles just fine, so I can probably use them instead.

khyperia commented on crates/rustc_codegen_spirv/src/abi.rs:
```diff
@@ -768,6 +773,92 @@ fn trans_intrinsic_type<'tcx>(
+                TyKind::Float(FloatTy::F32) => SpirvType::Float(32).def(span, cx),
+                TyKind::Float(FloatTy::F64) => SpirvType::Float(64).def(span, cx),
+                _ => {
+                    cx.tcx.sess.err("Invalid sampled type to `Image`.");
```
I think passing in the span would be good here?

khyperia commented on crates/spirv-builder/src/lib.rs:
```diff
@@ -204,7 +204,7 @@ fn invoke_rustc(builder: &SpirvBuilder, multimodule: bool) -> Result<PathBuf, Sp
 
     let rustflags = format!(
-        "-Z codegen-backend={} -Z symbol-mangling-version=v0{}",
+        "-Z codegen-backend={} {}",
```
Wait, why was this done? I probably missed a discussion around this, but I couldn't find anything in the review comments here.

XAMPPRocky replied:
So, it was required for a previous iteration of the PR, however my most recent changes removed the parts of `Image` that required the newer symbol mangling version, so this is no longer needed.

khyperia commented on crates/spirv-std/shared/Cargo.toml:
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "spirv-types"
+version = "0.4.0-alpha.3"
+authors = ["Erin Power <erin.power@embark-studios.com>"]
```
Nit: `Embark <opensource@embark-studios.com>`. Also `version = "0.4.0-alpha.5"`, and `repository = "https://github.com/EmbarkStudios/rust-gpu"`

khyperia commented on crates/spirv-std/src/image.rs:
```diff
@@ -0,0 +1,713 @@
+                "%sampler = OpLoad typeof*{2} {2}",
+                "%coord = OpLoad typeof*{3} {3}",
+                "%sampledImage = OpSampledImage %typeSampledImage %image %sampler",
+                "%result = OpImageSampleImplicitLod typeof*{0} %sampledImage %coord",
```
From the spec:

> OpImageSampleImplicitLod
> The MS operand of the underlying OpTypeImage must be 0.

Might be nice to encode that, as well as all the other restrictions here, but that can be done in a follow-up.

khyperia reviewed (commented):
Looks roughly good to me! The extra crate is a little annoying, but it works, so sure I guess. I think I might prefer nuking the old image system at the same time, having duplication for even a short amount of time is a big maintenance burden and causes confusion.

XAMPPRocky commented:
Alright, I've implemented all of the feedback.

> I think I might prefer nuking the old image system at the same time, having duplication for even a short amount of time is a big maintenance burden and causes confusion.

Well if we nuke the old system, we can't upgrade Ark until C-like enums are stabilised or we change the parts of Ark to not use reflection for entry points. So I do think it's worth keeping them both for a short time, I've also added a deprecated message if you have the const generics feature enabled, so if you can use it, it warns to move to the new type.

khyperia commented on crates/rustc_codegen_spirv/src/symbols.rs:
```diff
@@ -430,7 +432,9 @@ pub(crate) fn parse_attrs_for_checking<'a>(
             .chain(args.into_iter().map(move |ref arg| {
                 let span = arg.span();
-                let parsed_attr = if arg.has_name(sym.image_type) {
+                let parsed_attr = if arg.has_name(sym.generic_image_type) {
+                    SpirvAttribute::IntrinsicType(IntrinsicType::GenericImageType)
```
Is there a reason this isn't in the `custom_attributes` array? Seems like that's the proper place for it.

XAMPPRocky replied:
Well it's not how the `Image` was implemented, and this was just adapting from how that was implemented.

khyperia replied:
That's because `image` has arguments, `generic_image` does not - `symbols.rs` is organized by syntax, not by semantics.

khyperia commented on crates/spirv-builder/src/lib.rs:
```diff
@@ -204,7 +204,7 @@ fn invoke_rustc(builder: &SpirvBuilder, multimodule: bool) -> Result<PathBuf, Sp
 
     let rustflags = format!(
-        "-Z codegen-backend={} -Z symbol-mangling-version=v0{}",
+        "-Z codegen-backend={} -Zsymbol-mangling-version=v0 {}",
```
Nit: This will cause a double space if `llvm_args` has a value, and a trailing space if `llvm_args` is empty. How it was before was correct.

khyperia commented on tests/ui/image/sample_depth_reference/sample.rs:
```diff
@@ -1,19 +1,20 @@
     *output = image.sample_depth_reference(*sampler, v2, 1.0);
     *output += image_array.sample_depth_reference(*sampler, v3, 1.0);
-    *output += cubemap.sample_depth_reference(*sampler, v3, 1.0);
+    let sample: f32 = cubemap.sample_depth_reference(*sampler, v3, 1.0);
+    *output += sample;
```
Nit: Is there a reason for this change?

khyperia reviewed (approved):
LGTM pending nits.

Just realized, do we want to export common type aliases? Say, for all the types we already have explicitly defined, e.g. `pub type Image2D = Image!(2d, type=f32, sampled)` - might be nice for both examples for users of how to use the macro, plus saves users having to figure out e.g. what the heck `sampled` means and if they want it or not.

khyperia commented:
Oh yeah, also adding docs to https://embarkstudios.github.io/rust-gpu/book/attributes.html

XAMPPRocky commented:
> Oh yeah, also adding docs to the rust-gpu book on the macro syntax would be good, but that can be done in a follow-up if you'd like (file an issue if so).

The syntax is documented on the macro itself so when you visit the `Image!` API docs you'll see it there.

khyperia commented:
> The syntax is documented on the macro itself so when you visit the `Image!` API docs you'll see it there.

People look at the book instead of the API docs when trying to figure out language constructs (where language = DSL in this case), e.g. `u32` is documented both here https://doc.rust-lang.org/stable/std/primitive.u32.html and here https://doc.rust-lang.org/book/ch03-02-data-types.html . I think it'd make rust-gpu more accessible and easier to use if documentation was available in both places - something as core to graphics programming as images should be documented in the rust-gpu book, especially if it's not "normal" rust types.

XAMPPRocky commented on crates/spirv-std/src/lib.rs:
```diff
@@ -68,22 +68,28 @@
 pub mod scalar;
 pub(crate) mod sealed;
-mod textures;
+pub mod textures;
```
```suggestion
mod textures;
```

khyperia commented on crates/rustc_codegen_spirv/src/symbols.rs:
```diff
@@ -317,13 +317,21 @@ impl Symbols {
             ("invariant", SpirvAttribute::Invariant),
+            (
+                "generic_image",
+                SpirvAttribute::IntrinsicType(IntrinsicType::SampledImage),
```
huh?

XAMPPRocky replied:
```suggestion
                SpirvAttribute::IntrinsicType(IntrinsicType::GenericImageType),
```

XAMPPRocky commented on crates/rustc_codegen_spirv/src/symbols.rs:
```diff
@@ -317,13 +317,21 @@ impl Symbols {
             ("flat", SpirvAttribute::Flat),
             ("invariant", SpirvAttribute::Invariant),
+            (
+                "generic_image",
```
```suggestion
                "generic_image_type",
```

khyperia replied:
wait, now there are two of them?
